pkg/controller/persistentvolume/controller_base.go

```diff
@@ -0,0 +1,361 @@
+				claim = c
+			}
+		}
+	}
+
+	// sync the claim when its volume is deleted. Explicitly syncing the claim
+	// here in response to volume deletion prevents the claim from waiting until
+	// the next sync period for its Lost status.
+	if claim != nil {
+		err := ctrl.syncClaim(claim)
+		if err != nil {
+			if errors.IsConflict(err) {
+				// Version conflict error happens quite often and the controller
+				// recovers from it easily.
+				glog.V(3).Infof("PersistentVolumeController could not update volume %q from deleteClaim handler: %+v", claimToClaimKey(claim), err)
+			} else {
+				glog.Errorf("PersistentVolumeController could not update volume %q from deleteClaim handler: %+v", claimToClaimKey(claim), err)
+			}
+		}
+	}
+}
+
+// addClaim is callback from framework.Controller watching PersistentVolumeClaim
+// events.
+func (ctrl *PersistentVolumeController) addClaim(obj interface{}) {
+	if !ctrl.isFullySynced() {
+		return
+	}
+
+	claim, ok := obj.(*api.PersistentVolumeClaim)
+	if !ok {
+		glog.Errorf("Expected PersistentVolumeClaim but handler received %+v", obj)
+		return
+	}
+	if err := ctrl.syncClaim(claim); err != nil {
+		if errors.IsConflict(err) {
+			// Version conflict error happens quite often and the controller
+			// recovers from it easily.
+			glog.V(3).Infof("PersistentVolumeController could not add claim %q: %+v", claimToClaimKey(claim), err)
+		} else {
+			glog.Errorf("PersistentVolumeController could not add claim %q: %+v", claimToClaimKey(claim), err)
+		}
+	}
+}
+
+// updateClaim is callback from framework.Controller watching PersistentVolumeClaim
+// events.
+func (ctrl *PersistentVolumeController) updateClaim(oldObj, newObj interface{}) {
+	if !ctrl.isFullySynced() {
+		return
+	}
+
+	newClaim, ok := newObj.(*api.PersistentVolumeClaim)
+	if !ok {
+		glog.Errorf("Expected PersistentVolumeClaim but handler received %+v", newObj)
+		return
+	}
+	if err := ctrl.syncClaim(newClaim); err != nil {
+		if errors.IsConflict(err) {
+			// Version conflict error happens quite often and the controller
+			// recovers from it easily.
+			glog.V(3).Infof("PersistentVolumeController could not update claim %q: %+v", claimToClaimKey(newClaim), err)
+		} else {
+			glog.Errorf("PersistentVolumeController could not update claim %q: %+v", claimToClaimKey(newClaim), err)
+		}
+	}
+}
+
+// deleteClaim is callback from framework.Controller watching PersistentVolumeClaim
+// events.
+func (ctrl *PersistentVolumeController) deleteClaim(obj interface{}) {
+	if !ctrl.isFullySynced() {
+		return
+	}
+
+	var volume *api.PersistentVolume
+	if pvc, ok := obj.(*api.PersistentVolumeClaim); ok {
+		if pvObj, exists, _ := ctrl.volumes.store.GetByKey(pvc.Spec.VolumeName); exists {
+			if pv, ok := pvObj.(*api.PersistentVolume); ok {
+				volume = pv
+			}
+		}
+	}
+	if unk, ok := obj.(cache.DeletedFinalStateUnknown); ok && unk.Obj != nil {
+		if pv, ok := unk.Obj.(*api.PersistentVolume); ok {
```

Comment:
Seems like a bug.  I'm guessing this cast always fails and things just continue on their merry way, but this seems wrong.  Note: we all missed this, only @jsafrane noticed it.